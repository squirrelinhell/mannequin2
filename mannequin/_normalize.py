
import numpy as np

class RunningMean(object):
    def __init__(self, shape=(), *, horizon=None):
        mean = np.zeros(shape, dtype=np.float64)
        total_weight = 0.0

        def update(values, *, weight=1.0):
            nonlocal mean, total_weight

            values = np.asarray(values, dtype=np.float64)
            values = values.reshape(shape)

            weight = float(weight)
            assert weight >= 0.0

            if horizon is not None:
                total_weight *= np.power(
                    1.0 - 1.0 / float(horizon),
                    weight
                )

            total_weight += weight
            mean += (weight / total_weight) * (values - mean)

        self.get = lambda: np.array(mean)
        self.update = update

    def __call__(self, values, **kwargs):
        self.update(values, **kwargs)
        return self.get()

class RunningNormalize(object):
    def __init__(self, shape=(), *, horizon=None):
        r_mean = RunningMean(shape=shape, horizon=horizon)
        r_var = RunningMean(shape=shape, horizon=horizon)
        n_samples = 0

        def update(value):
            nonlocal n_samples

            value = np.array(value, dtype=np.float64)
            value = value.reshape((-1,) + shape)

            d1 = value - r_mean.get()
            d2 = value - r_mean(np.mean(value, axis=0))

            if n_samples >= 1:
                r_var.update(np.mean(np.multiply(d1, d2), axis=0))
            elif len(value) >= 2:
                weight = (len(value) - 1.0) / len(value)
                r_var.update(
                    np.mean(np.multiply(d1, d2), axis=0) / weight,
                    weight=weight
                )

            n_samples += len(value)

        def apply(value):
            if n_samples < 2:
                return np.zeros_like(value, dtype=np.float64)
            return (
                (value - r_mean.get())
                / np.maximum(1e-8, np.sqrt(r_var.get()))
            )

        def get_var():
            if n_samples < 2:
                return np.ones(shape, dtype=np.float64)
            return r_var.get()

        self.update = update
        self.apply = apply
        self.get_mean = lambda: r_mean.get()
        self.get_var = get_var
        self.get_std = lambda: np.sqrt(get_var())

    def __call__(self, value):
        self.update(value)
        return self.apply(value)
