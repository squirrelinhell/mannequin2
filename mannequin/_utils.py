
import numpy as np

def endswith(a, b):
    if len(a) < len(b):
        return False
    return a[len(a)-len(b):] == b

def discounted(values, *, horizon):
    step = 1.0 / float(horizon)
    assert (step > 1e-6) and (step < 1.0)

    values = np.array(values, dtype=np.float64)
    rew_sum = 0.0

    for i in range(len(values)-1, -1, -1):
        rew_sum += step * (values[i] - rew_sum)
        values[i] = rew_sum

    return values

def bar(value, max_value=100.0, *, length=50):
    value = float(value)
    max_value = abs(float(max_value))
    length = max(1, int(length))

    if value >= 0.0:
        bar = max(0.0, min(1.0, value / max_value))
        bar, frac = divmod(int(round(bar * length * 8)), 8)
        bar = chr(0x2588) * bar
        if frac >= 1:
            bar += chr(0x2590 - frac)
        bar += " " * length
        bar = bar[:length]
    else:
        bar = max(0.0, min(1.0, abs(value) / max_value))
        bar, frac = divmod(int(round(bar * length * 2)), 2)
        bar = chr(0x2501) * bar
        if frac >= 1:
            bar = chr(0x257A) + bar
        bar = " " * length + bar
        bar = bar[-length:]

    fmt_len = len("-%.2f" % max_value)
    fmt = ("%%%d.2f " % fmt_len) % value
    return fmt[0:fmt_len+1] + "[" + bar + "]"

def plot_function_2d(function, *,
        xlim=(-2.0, 2.0),
        ylim=(-2.0, 2.0),
        vlim=(None, None),
        grid=20,
        file_name=None):
    import matplotlib.pyplot as plt
    coords = (np.mgrid[0:grid+1,0:grid+1].reshape(2,-1).T / grid
        * [xlim[1] - xlim[0], ylim[1] - ylim[0]] + [xlim[0], ylim[0]])
    values = np.array([float(function(*c)) for c in coords])
    values = values.reshape((grid+1, grid+1))
    plt.clf()
    plt.imshow(
        values.T[::-1,:], vmin=vlim[0], vmax=vlim[1],
        zorder=0, aspect="auto",
        cmap="inferno", interpolation="bicubic",
        extent=[xlim[0], xlim[1], xlim[0], ylim[1]]
    )
    plt.colorbar()
    if file_name is None:
        plt.ion()
        plt.show()
        plt.pause(0.0001)
    else:
        plt.gcf().set_size_inches(10, 8)
        plt.gcf().savefig(file_name, dpi=100)
