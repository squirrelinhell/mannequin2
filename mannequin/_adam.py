
class BaseTwoMoments:
    def __init__(self, value, update_rule, *,
            horizon=10,
            var_horizon=100,
            print_norm=False):
        import numpy as np
        import os
        from mannequin import RunningMean

        value = np.asarray(value, dtype=np.float64)
        value.setflags(write=False)

        running_mean = RunningMean(value.shape, horizon=horizon)
        running_var = RunningMean(value.shape, horizon=var_horizon)

        def norm(v):
            return np.sqrt(np.sum(np.square(v)))

        def apply_gradient(grad):
            nonlocal value

            grad = np.asarray(grad, dtype=np.float64)
            assert grad.shape == value.shape

            running_mean.update(grad)
            running_var.update(np.square(grad))

            add = update_rule(running_mean.get(), running_var.get())

            if print_norm:
                print("Update norm: %10.4f" % norm(add))

            value = value + add
            value.setflags(write=False)

        self.get_value = lambda: value
        self.apply_gradient = apply_gradient

class Adam(BaseTwoMoments):
    def __init__(self, value, *, lr, epsilon=1e-8, **params):
        import numpy as np
        lr = float(lr)
        epsilon = float(epsilon)

        def update_rule(mean, var):
            return lr * (mean / (epsilon + np.sqrt(var)))

        super().__init__(value, update_rule, **params)

class Adams(BaseTwoMoments):
    def __init__(self, value, *, lr, epsilon=1e-8, **params):
        lr = float(lr)
        epsilon = float(epsilon)

        def update_rule(mean, var):
            return lr * (mean / (epsilon + var))

        super().__init__(value, update_rule, **params)
