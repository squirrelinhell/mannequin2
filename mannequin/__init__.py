
from ._adam import *
from ._trajectory import *
from ._utils import *

version = (2, 2)
