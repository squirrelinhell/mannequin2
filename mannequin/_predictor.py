
import numpy as np

from mannequin import Adam, Trajectory, RunningNormalize
from mannequin.basicnet import Input, Affine, Tanh

class SimplePredictor(object):
    def __init__(self, in_size=1, out_size=1, *,
            hid_layers=2, hid_size=64, activation=Tanh,
            classifier=False, normalize_inputs=False):
        in_size = int(in_size)
        out_size = int(out_size)
        hid_size = int(hid_size)

        # Build a simple neural network
        model = Input(in_size)
        for _ in range(hid_layers):
            model = activation(Affine(model, hid_size))
        model = Affine(model, out_size, init=0.1)

        # Use default parameter initialization
        opt = Adam(model.get_params(), horizon=10)

        if normalize_inputs:
            normalize = RunningNormalize(shape=(in_size,))

        def softmax(v):
            v = v.T
            v = np.exp(v - np.amax(v, axis=0))
            v /= np.sum(v, axis=0)
            return v.T

        def predict(inputs):
            inputs = np.asarray(inputs)
            if in_size == 1 and inputs.shape[-1:] != (1,):
                inputs = inputs.reshape(inputs.shape + (1,))
            if normalize_inputs:
                inputs = normalize.apply(inputs)
            outs, _ = model.evaluate(inputs)
            if classifier:
                return softmax(outs)
            elif out_size == 1:
                return outs.T[0].T
            else:
                return outs

        def sgd_step(traj, labels=None, *, lr=0.01):
            if labels is not None:
                traj = Trajectory(traj, labels)
                del labels
            assert isinstance(traj, Trajectory)
            inps = traj.o.reshape((-1, in_size))
            lbls = traj.a.reshape((-1, out_size))
            inps = normalize(inps) if normalize_inputs else inps
            outs, backprop = model.evaluate(inps)
            if classifier:
                outs = softmax(outs) + outs * 0.01
            grad = np.multiply((lbls - outs).T, traj.r).T
            opt.apply_gradient(backprop(grad), lr=lr)
            model.load_params(opt.get_value())

        self.predict = predict
        self.sgd_step = sgd_step

    def __call__(self, inputs):
        return self.predict(inputs)
