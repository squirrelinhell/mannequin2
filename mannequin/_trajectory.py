
import numpy as np

import mannequin

class Trajectory(object):
    def __init__(self, observations, actions, rewards=None):
        # Standarize observations
        if isinstance(observations[0], (int, np.integer)):
            observations = np.asarray(observations, dtype=np.int32)
        else:
            observations = np.asarray(observations, dtype=np.float32)
        length = len(observations)
        if length < 1:
            raise ValueError("Cannot create empty trajectory")

        # Standarize actions
        if isinstance(actions[0], (int, np.integer)):
            actions = np.asarray(actions, dtype=np.int32)
        else:
            actions = np.asarray(actions, dtype=np.float32)
        if len(actions) != length:
            raise ValueError("Actions don't match observations")

        # Standarize rewards
        if rewards is None:
            rewards = np.ones(length, dtype=np.float32)
        else:
            rewards = np.asarray(rewards, dtype=np.float32)
            rewards = np.broadcast_to(rewards, (length,))

        # Make all arrays read-only
        observations.setflags(write=False)
        actions.setflags(write=False)
        rewards.setflags(write=False)

        def modified(**kwargs):
            data = {
                "observations": observations,
                "actions": actions,
                "rewards": rewards
            }
            for v in kwargs:
                data[v] = kwargs[v](data[v])
            return Trajectory(**data)

        def discounted(*, horizon):
            return Trajectory(
                observations=observations,
                actions=actions,
                rewards=mannequin.discounted(rewards, horizon=horizon)
            )

        # Public methods
        self.get_length = lambda: length
        self.get_observations = lambda: observations[:]
        self.get_actions = lambda: actions[:]
        self.get_rewards = lambda: rewards[:]
        self.modified = modified
        self.discounted = discounted

    def __getattribute__(self, name):
        if name == "observations"[:len(name)]:
            return self.get_observations()
        if name == "actions"[:len(name)]:
            return self.get_actions()
        if name == "rewards"[:len(name)]:
            return self.get_rewards()
        return super().__getattribute__(name)

    def __setattr__(self, name, value):
        if name[:1] in ("o", "a", "r"):
            raise ValueError("Field '%s' is read-only" % name)
        return super().__setattr__(name, value)

    def __getitem__(self, idx):
        o = self.get_observations()[idx]
        a = self.get_actions()[idx]
        r = self.get_rewards()[idx]
        if len(r.shape) >= 1:
            return Trajectory(o, a, r)
        else:
            return Trajectory([o], [a], [r])

    def __len__(self):
        return self.get_length()

    def __str__(self):
        return "<trajectory: %d x %s -> %s>" % (
            self.get_length(),
            self.get_observations().shape[1:],
            self.get_actions().shape[1:]
        )

    def joined(*ts):
        if len(ts) <= 1:
            ts = ts[0]
        return Trajectory(
            np.concatenate([t.get_observations() for t in ts], axis=0),
            np.concatenate([t.get_actions() for t in ts], axis=0),
            np.concatenate([t.get_rewards() for t in ts], axis=0)
        )
