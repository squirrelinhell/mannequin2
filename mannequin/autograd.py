
import inspect
import autograd
import autograd.numpy as np

from mannequin.basicnet import Layer, Input, equalized_columns

class AutogradLayer(Layer):
    def __init__(self, inner, *, f, params=None, **layer_args):
        f_args_names = inspect.getfullargspec(f).kwonlyargs
        df = autograd.grad(lambda a, ka, g: np.sum(f(*a, **ka) * g))

        def evaluate(inps, **args):
            f_args = {}
            for a in f_args_names:
                if not a in args:
                    raise TypeError("evaluate() missing"
                        + " required argument: '%s'" % a)
                f_args[a] = args[a]
                del args[a]
            inps = np.asarray(inps, dtype=np.float32)
            inps, inner_backprop = inner.evaluate(inps, **args)
            assert inps.shape == (*inps.shape[:-1], inner.n_outputs)
            if params is None:
                outs = f(inps, **f_args)
            else:
                outs = f(inps, params, **f_args)
            outs = np.asarray(outs, dtype=np.float32)
            assert outs.shape == (*inps.shape[:-1], self.n_outputs)
            def backprop(grad):
                nonlocal inps
                inps = np.reshape(inps, (-1, inner.n_outputs))
                grad = np.asarray(grad, dtype=np.float32)
                grad = np.reshape(grad, (-1, self.n_outputs))
                if params is None:
                    i_grad, = df((inps,), f_args, grad)
                    return inner_backprop(i_grad)
                else:
                    i_grad, p_grad = df((inps, params), f_args, grad)
                    return np.concatenate((
                        inner_backprop(i_grad),
                        p_grad.reshape(-1) / len(grad)
                    ), axis=0)
            return outs, backprop

        super().__init__(inner, evaluate=evaluate,
            params=params, **layer_args)

def Linear(inner, n_outputs, *, init=equalized_columns):
    params = init(inner.n_outputs, n_outputs).astype(np.float32)
    multiplier = 1.0 / np.sqrt(float(inner.n_outputs))
    return AutogradLayer(inner, n_outputs=n_outputs,
        f=lambda i, w: np.dot(i, w) * multiplier, params=params)

def Bias(inner, *, init=np.zeros):
    params = init(inner.n_outputs).astype(np.float32)
    return AutogradLayer(inner, f=np.add, params=params)

def LReLU(inner, leak=0.1):
    leak = float(leak)
    return AutogradLayer(inner,
        f=lambda i: np.maximum(0.0, i) + leak * np.minimum(0.0, i))

def Tanh(inner):
    return AutogradLayer(inner, f=np.tanh)

def Affine(*args, **kwargs):
    return Bias(Linear(*args, **kwargs))
