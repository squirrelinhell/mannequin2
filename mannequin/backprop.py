
import numpy as np

def autograd(f):
    import autograd
    import autograd.numpy as np
    df = autograd.grad(lambda a, ka, g: np.sum(f(*a, **ka) * g))
    return lambda *a, **ka: (f(*a, **ka), lambda g: df(a, ka, g))

def add(a, b):
    ash, bsh = a.shape, b.shape
    if ash == bsh:
        return a + b, lambda g: (g, g)
    elif len(ash) > len(bsh) and ash[len(ash)-len(bsh):] == bsh:
        return a + b, lambda g: (
            g,
            np.sum(g, axis=tuple(range(len(ash)-len(bsh))))
        )
    else:
        raise ValueError("Invalid shapes: %s, %s" % (a.shape, b.shape))

def multiply(a, b):
    assert a.shape == b.shape
    return a * b, lambda g: (g * b, g * a)

def matmul(a, b):
    assert len(a.shape) in (1, 2)
    assert len(b.shape) == 2
    return np.matmul(a, b), lambda g: (
        np.matmul(g, b.T),
        np.matmul(a.T, g) if len(a.shape) >= 2 else a[:,None] * g
    )

def relu(a, *, leak=0.0):
    multiplier = np.ones(a.shape)
    multiplier[a < 0.0] = float(leak)
    return a * multiplier, lambda g: (g * multiplier,)

def tanh(a):
    a = np.tanh(a)
    return a, lambda g: (g * (1.0 - np.square(a)),)
