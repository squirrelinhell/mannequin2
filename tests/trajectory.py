
from test_setup import timer
from mannequin import Trajectory

t = Trajectory(
    [[1, 2], [3, 4], [5, 6]],
    [10, 11, 12],
    [20, 21, 22]
)

print(t)
print(t[1:])
print(t[1])

t = t[::-1]
print(t.o)
print(t.a)
print(t.r)

t = t.joined(t[0])
print(t.r)

t = t.modified(rewards=sum)
print(t.r)

assert timer() < 0.01
