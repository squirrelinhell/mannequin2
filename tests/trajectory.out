<trajectory: 3 x (2,) -> (1,)>
<trajectory: 2 x (2,) -> (1,)>
<trajectory: 1 x (2,) -> (1,)>
[[ 5.  6.]
 [ 3.  4.]
 [ 1.  2.]]
[[ 12.]
 [ 11.]
 [ 10.]]
[ 22.  21.  20.]
[ 22.  21.  20.  22.]
[ 85.  85.  85.  85.]
